how to use `System.Collections.Immutable` v`8.0.0` with `Microsoft.CodeAnalysis.Testing.ReferenceAssemblies.Net.Net60`

our code-analyzer-tests use `Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.MSTest`. We now had to upgrade `System.Collections.Immutable`  to `8.0.0` for all Assemblies. Whenever we now want to use these assemblies to run within tests (because we have tests specifically for Classes inside our assemblies) we get an Exception:
```txt
// error CS1705: Assembly 'Komsa' with identity 'Komsa, Version=2.23.1.0, Culture=neutral, PublicKeyToken=null' uses 'System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'System.Collections.Immutable' with identity 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
DiagnosticResult.CompilerError("CS1705").WithArguments("Komsa", "Komsa, Version=2.23.1.0, Culture=neutral, PublicKeyToken=null", "System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", "System.Collections.Immutable", "System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"),
```

adding
```cs
new Verifier.Test()
{
  TestState =
  {
    AdditionalReferences = { typeof(ImmutableArray).Assembly  },
  }
}
```
results in the following:
```txt
Actual diagnostic:
    // /0/Test0.cs(13,5): error CS0433: The type 'ImmutableArray<T>' exists in both 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
DiagnosticResult.CompilerError("CS0433").WithSpan(13, 5, 13, 27).WithArguments("System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", "System.Collections.Immutable.ImmutableArray<T>", "System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"),
```

adding the NugetPackage like this does not change anything (in either above case):
```cs
new Verifier.Test()
{
  ReferenceAssemblies = ReferenceAssemblies.Net.Net60 // used always
    .AddPackages([new PackageIdentity("System.Collections.Immutable", "8.0.0")]) // new
}
```

Our verifiers are essentially a "thin wrapper" around `CodeFixVerifier<TAnalyzer, TCodeFix, TTest, TVerifier>` and Verifier.Test around `CSharpCodeFixTest<TAnalyzer, TCodeFix, TVerifier>`, to allow for more testing and better migrating from our old tests, and as a workaround for other bugs in the past (ie tests not being able to handle 2-space indents correctly). We do not interfere with Roslyn in this subclass, we "only extend/append" (editorconfig-spacing) on what the base implementation provides. Thus I don't see us being the reason for "something not working"

--- a/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
+++ b/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
@@ -460,6 +460,53 @@ namespace Microsoft.CodeAnalysis.Testing
                     }
                 }
 
+                // Prefer newer assemblies when more than one have the same name
+                if (ReferenceAssemblyPackage is not null)
+                {
+                    var comparer = new FrameworkPrecedenceSorter(DefaultFrameworkNameProvider.Instance, allEquivalent: false);
+                    var assembliesByName = resolvedAssemblies.GroupBy(Path.GetFileNameWithoutExtension, StringComparer.OrdinalIgnoreCase);
+                    var assembliesToRemove = new List<string>();
+                    foreach (var assemblyNameGroup in assembliesByName)
+                    {
+                        var assembliesByPrecedence = assemblyNameGroup.OrderBy(GetFrameworkNameFromPath, comparer).ToArray();
+                        for (var i = 1; i < assembliesByPrecedence.Length; i++)
+                        {
+                            // We want to keep the last reference listed for the most recent supported target framework.
+                            // Typically, if more than one item has the most recent supported target framework, it will
+                            // be a case where the reference assembly package provides the assembly and a newer version
+                            // is provided explicitly. For example:
+                            //
+                            // Microsoft.NETCore.App.Ref 6.0.0 provides System.Collections.Immutable in the net6.0 folder
+                            // System.Collections.Immutable 8.0.0 provides System.Collections.Immutable in the net6.0 folder
+                            //
+                            // In this example, the Microsoft.NETCore.App.Ref package is resolved first, so by taking
+                            // the last net6.0 assembly, we ensure the assembly from System.Collections.Immutable 8.0.0
+                            // is resolved.
+                            if (comparer.Compare(GetFrameworkNameFromPath(assembliesByPrecedence[0]), GetFrameworkNameFromPath(assembliesByPrecedence[i])) == 0)
+                            {
+                                assembliesToRemove.Add(assembliesByPrecedence[i - 1]);
+                            }
+                            else
+                            {
+                                assembliesToRemove.Add(assembliesByPrecedence[i]);
+                            }
+                        }
+
+                        static NuGetFramework GetFrameworkNameFromPath(string path)
+                        {
+                            var frameworkFolder = Path.GetFileName(Path.GetDirectoryName(path));
+                            if (frameworkFolder is null)
+                            {
+                                return NuGetFramework.UnsupportedFramework;
+                            }
+
+                            return NuGetFramework.ParseFolder(frameworkFolder);
+                        }
+                    }
+
+                    resolvedAssemblies.ExceptWith(assembliesToRemove);
+                }
+
                 // Add the facade assemblies
                 if (ReferenceAssemblyPackage is not null)
                 {

--- a/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
+++ b/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
@@ -465,10 +465,17 @@ namespace Microsoft.CodeAnalysis.Testing
                 {
                     var comparer = new FrameworkPrecedenceSorter(DefaultFrameworkNameProvider.Instance, allEquivalent: false);
                     var assembliesByName = resolvedAssemblies.GroupBy(Path.GetFileNameWithoutExtension, StringComparer.OrdinalIgnoreCase);
+
+                    // Keep track of assemblies to remove from resolvedAssemblies. Defer the actual removal to the end
+                    // of this block for ease in future debugging scenarios.
                     var assembliesToRemove = new List<string>();
                     foreach (var assemblyNameGroup in assembliesByName)
                     {
-                        var assembliesByPrecedence = assemblyNameGroup.OrderBy(GetFrameworkNameFromPath, comparer).ThenByDescending(GetFrameworkNameFromPath, new NuGetFrameworkSorter()).ToArray();
+                        var assembliesByPrecedence = assemblyNameGroup
+                            .Select(static name => (name, framework: GetFrameworkNameFromPath(name)))
+                            .OrderBy(static x => x.framework, comparer)
+                            .ThenByDescending(static x => x.framework, new NuGetFrameworkSorter())
+                            .ToArray();
                         for (var i = 1; i < assembliesByPrecedence.Length; i++)
                         {
                             // We want to keep the last reference listed for the most recent supported target framework.
@@ -482,13 +489,13 @@ namespace Microsoft.CodeAnalysis.Testing
                             // In this example, the Microsoft.NETCore.App.Ref package is resolved first, so by taking
                             // the last net6.0 assembly, we ensure the assembly from System.Collections.Immutable 8.0.0
                             // is resolved.
-                            if (Equals(GetFrameworkNameFromPath(assembliesByPrecedence[0]), GetFrameworkNameFromPath(assembliesByPrecedence[i])))
+                            if (Equals(assembliesByPrecedence[0].framework, assembliesByPrecedence[i].framework))
                             {
-                                assembliesToRemove.Add(assembliesByPrecedence[i - 1]);
+                                assembliesToRemove.Add(assembliesByPrecedence[i - 1].name);
                             }
                             else
                             {
-                                assembliesToRemove.Add(assembliesByPrecedence[i]);
+                                assembliesToRemove.Add(assembliesByPrecedence[i].name);
                             }
                         }
 

--- a/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
+++ b/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
@@ -460,16 +460,6 @@ namespace Microsoft.CodeAnalysis.Testing
                     }
                 }
 
-                // Prefer assemblies from the reference assembly package to ones otherwise provided
-                if (ReferenceAssemblyPackage is not null)
-                {
-                    var referenceAssemblies = new HashSet<string>(resolvedAssemblies.Where(resolved => resolved.StartsWith(referenceAssemblyInstalledPath!)));
-
-                    // Suppression due to https://github.com/dotnet/roslyn/issues/44735
-                    var referenceAssemblyNames = new HashSet<string>(referenceAssemblies.Select((Func<string, string>)Path.GetFileNameWithoutExtension!));
-                    resolvedAssemblies.RemoveWhere(resolved => referenceAssemblyNames.Contains(Path.GetFileNameWithoutExtension(resolved)) && !referenceAssemblies.Contains(resolved));
-                }
-
                 // Add the facade assemblies
                 if (ReferenceAssemblyPackage is not null)
                 {

--- a/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
+++ b/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs
@@ -468,7 +468,7 @@ namespace Microsoft.CodeAnalysis.Testing
                     var assembliesToRemove = new List<string>();
                     foreach (var assemblyNameGroup in assembliesByName)
                     {
-                        var assembliesByPrecedence = assemblyNameGroup.OrderBy(GetFrameworkNameFromPath, comparer).ToArray();
+                        var assembliesByPrecedence = assemblyNameGroup.OrderBy(GetFrameworkNameFromPath, comparer).ThenByDescending(GetFrameworkNameFromPath, new NuGetFrameworkSorter()).ToArray();
                         for (var i = 1; i < assembliesByPrecedence.Length; i++)
                         {
                             // We want to keep the last reference listed for the most recent supported target framework.
@@ -482,7 +482,7 @@ namespace Microsoft.CodeAnalysis.Testing
                             // In this example, the Microsoft.NETCore.App.Ref package is resolved first, so by taking
                             // the last net6.0 assembly, we ensure the assembly from System.Collections.Immutable 8.0.0
                             // is resolved.
-                            if (comparer.Compare(GetFrameworkNameFromPath(assembliesByPrecedence[0]), GetFrameworkNameFromPath(assembliesByPrecedence[i])) == 0)
+                            if (Equals(GetFrameworkNameFromPath(assembliesByPrecedence[0]), GetFrameworkNameFromPath(assembliesByPrecedence[i])))
                             {
                                 assembliesToRemove.Add(assembliesByPrecedence[i - 1]);
                             }

--- a/tests/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing.UnitTests/MetadataReferenceTests.cs
+++ b/tests/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing.UnitTests/MetadataReferenceTests.cs
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
 using Xunit;
@@ -646,6 +647,37 @@ class TestClass {
             {
                 TestCode = testCode,
                 ReferenceAssemblies = ReferenceAssembliesForTargetFramework(targetFramework),
+            }.RunAsync();
+        }
+
+        [Theory]
+        [InlineData("net462")]
+        [InlineData("net47")]
+        [InlineData("net471")]
+        [InlineData("net472")]
+        [InlineData("net48")]
+#if !(NETCOREAPP1_1 || NET46)
+        [InlineData("net6.0")]
+        [InlineData("net7.0")]
+        [InlineData("net8.0")]
+#endif
+        [InlineData("netstandard2.0")]
+        [InlineData("netstandard2.1")]
+        public async Task ResolveSystemCollectionsImmutable8(string targetFramework)
+        {
+            var testCode = @"
+using System.Collections.Frozen;
+
+class TestClass {
+  FrozenSet<int> TestMethod() => throw null;
+}
+";
+
+            await new CSharpTest()
+            {
+                TestCode = testCode,
+                ReferenceAssemblies = ReferenceAssembliesForTargetFramework(targetFramework)
+                    .AddPackages(ImmutableArray.Create(new PackageIdentity("System.Collections.Immutable", "8.0.0"))),
             }.RunAsync();
         }
 

